test/cmpd_dset.c

```diff
@@ -700,27 +964,126 @@ test_select_compound(char *fname, hid_t fapl, hid_t in_dxpl, unsigned set_fillva
         s3[i].e    = 8 * i + 7;
     }
 
-    /* Save a copy as the buffer may be clobbered due to H5Pset_modify_write_buf() */
     memcpy(save_s3, s3, sizeof(s3_t) * NX * NY);
 
+    /* Set buffer size accordingly */
     if (set_buf) {
-        ss1 = H5Tget_size(s1_tid);
+        /* ss1 is set already previously */
         ss2 = H5Tget_size(s3_tid);
         ss  = MAX(ss1, ss2) * NX * NY;
 
         if (H5Pset_buffer(dxpl, ss, NULL, NULL) < 0)
             goto error;
     }
 
-    /* Write to the dataset with s3_tid */
-    if (H5Dwrite(did, s3_tid, H5S_ALL, H5S_ALL, dxpl, s3) < 0)
+    /* Read from the dataset with mem type s3_tid */
+    if (H5Dread(did, s3_tid, H5S_ALL, H5S_ALL, dxpl, rbuf3) < 0)
+        goto error;
+
+    if (compare_s1_s3_data(save_s1, rbuf3) < 0)
         goto error;
 
-    if (H5Dread(did, s3_tid, H5S_ALL, H5S_ALL, dxpl, rbuf3) < 0)
+    if (H5Dclose(did) < 0)
+        goto error;
+
+    PASSED();
+
+    printf("    test_select_compound(): verify the CI window failure is fixed");
+
+    /* Create file type s7_tid */
+    if ((s7_tid = H5Tcreate(H5T_COMPOUND, sizeof(s7_t))) < 0)
+        goto error;
+
+    if (H5Tinsert(s7_tid, "a", HOFFSET(s7_t, a), H5T_NATIVE_INT) < 0 ||
+        H5Tinsert(s7_tid, "d", HOFFSET(s7_t, d), H5T_NATIVE_INT) < 0) 
+        goto error;
+
+    /* Initialize buffer with s7_t */
+    for (i = 0; i < NX * NY; i++) {
+        s7[i].a = (int32_t)(i * 2 + 1);
+        s7[i].d = (int32_t)(i * 2 + 2);
+    }
+    memcpy(save_s7, s7, sizeof(s7_t) * NX * NY);
+
+    /* Create mem type s8_tid */
+    if ((s8_tid = H5Tcreate(H5T_COMPOUND, sizeof(s8_t))) < 0)
+        goto error;
+
+    if (H5Tinsert(s8_tid, "a", HOFFSET(s8_t, a), H5T_NATIVE_LONG) < 0 ||
+        H5Tinsert(s8_tid, "b", HOFFSET(s8_t, b), H5T_NATIVE_LONG) < 0 ||
+        H5Tinsert(s8_tid, "c", HOFFSET(s8_t, c), H5T_NATIVE_LONG) < 0)
+        goto error;
+
+    /* Initialize buffer with s8_t */
+    for (i = 0; i < NX * NY; i++) {
+        s8[i].a = (int64_t)(i * 3 + 0);
+        s8[i].b = (int64_t)(i * 3 + 1);
+        s8[i].c = (int64_t)(i * 3 + 2);
+    }
+    memcpy(save_s8, s8, sizeof(s8_t) * NX * NY);
+
+    /* Set fill value accordingly */
+    if (set_fillvalue) {
+        if (H5Pset_fill_value(dcpl, s7_tid, &fillvalue) < 0)
+            goto error;
+    }
+
+    /* Set buffer size accordingly */
+    if (set_buf) {
+        ss1 = H5Tget_size(s7_tid);
+        ss2 = H5Tget_size(s8_tid);
+        ss  = MAX(ss1, ss2) * NX * NY;
+
+        if (H5Pset_buffer(dxpl, ss, NULL, NULL) < 0)
```

Comment:
Why are you doing this? The buffer needs to be smaller than the I/O size.

Reply:
The reason for doing this is because the default without setting H5Pset_buffer() is smaller than the I/O size.

For default without setting H5Pset_buffer():
--H5CX_get_max_temp_buf(&max_temp_buf) is 1048576
--type_info->request_nelmts =43690
--dset_info->nelmts = 200000
--the "for" loop in H5Dscatgath_write() will be triggered many times to fulfill the I/O requests

When setting H5Pset_buffer:
--H5CX_get_max_temp_buf(&max_temp_buf) is 4800000
--type_info->request_nelmts = 200000
--dset_info->nelmts = 200000
--the "for" loop in H5Dscatgath_write() will be triggered once

Reply:
The test is for the case when the buffer size is smaller than the I/O size. This will break selection I/O but that's not what we're testing.

Reply:
I'm surprised you're seeing the Linux failure (with the fix reverted) with the buffer big enough for the I/O.